src/mds/MDCache.cc

```diff
@@ -8441,9 +8439,19 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,
       if (dn->state_test(CDentry::STATE_PURGING))
 	return -CEPHFS_ENOENT;
 
+      CDentry::linkage_t *dnl = dn->get_projected_linkage();
+      // If an auth check was deferred before and the target inode is found
+      // not to exist now, do the auth check here if necessary.
+      if (want_auth && want_dentry && want_inode && depth == path.depth() - 1 &&
+	  dnl->is_null() && (r = maybe_request_forward_to_auth(mdr, cf, dn)) != 0)
```

Comment:
> Shouldn't this be `!dnl->is_null()`?

> yeah, I guess dentry linkage needs to be valid to take any sensible action on it so the linkage has to either be local or remote

`dnl->is_null()` here means that the target inode does not exist, so we need to give an auth tail dentry to the caller. But the tail dentry may not be auth here because we have canceled an auth check if `want_inode` is true. So we need to do an auth check for the tail dentry here. (here `dn` can be replaced by `curdir`)

Reply:
> If its not a null dentry it can be a primary or a remote dentry. If its a remote dentry, is the request (open(..., O_TRUNC) on hardlink) forwarded to the auth mds of the inode? Also, what if the dentry/inode is a replica?

If it is a remote dentry and the inode it links to is not auth(a replica) in current mds, the request will be forwarded to the mds with auth inode. The auth check and request forward for the target inode will be done at the tail of this function if the target inode exists.

Reply:
ACK. I'll have another look today. Seems to be going the right direction.